src/stan/lang/generator/write_begin_all_dims_row_maj_loop.hpp

```diff
@@ -0,0 +1,91 @@
+#ifndef STAN_LANG_GENERATOR_WRITE_BEGIN_ALL_DIMS_ROW_MAJ_LOOP_HPP
+#define STAN_LANG_GENERATOR_WRITE_BEGIN_ALL_DIMS_ROW_MAJ_LOOP_HPP
+
+#include <stan/lang/ast.hpp>
+#include <stan/lang/generator/constants.hpp>
+#include <stan/lang/generator/generate_expression.hpp>
+#include <stan/lang/generator/generate_indent.hpp>
+#include <ostream>
+#include <string>
+#include <vector>
+
+namespace stan {
+  namespace lang {
+
+    /**
+     * Generate the openings of a sequence of zero or more for loops
+     * corresponding to all dimensions of a variable, with the
+     * specified indentation level writing to the specified stream.
+     * If specified, declare named size_t variable for each dimension
+     * which avoids re-evaluation of size expression on each iteration.
+     *
+     * Indexing order is row major: array dims 1...N then row, col
+     *
+     * @param[in] var_decl variable declaration
+     * @param[in] declare_size_vars if true, generate size_t var decls
+     * @param[in] indent indentation level
+     * @param[in,out] o stream for generating
+     */
+
```

Comment:
So maybe we need column-major and row-major dimension extraction.  But with that, then this could be combined with the column major loop function.